zerver/lib/actions.py

```diff
@@ -463,18 +463,21 @@ def process_new_human_user(user_profile: UserProfile,
                 prereg_user.referred_by,
                 _("{user} accepted your invitation to join Zulip!").format(user=f"{user_profile.full_name} <`{user_profile.email}`>")
             )
-    # Mark any other PreregistrationUsers that are STATUS_ACTIVE as
+    # Mark any other PreregistrationUsers in the realm that are STATUS_ACTIVE as
     # inactive so we can keep track of the PreregistrationUser we
-    # actually used for analytics
-    if prereg_user is not None:
+    # actually used for analytics.
+    # In the special case of realm creation, there can be no additional PreregistrationUser
+    # for us to want to modify.
+    if prereg_user is not None and not realm_creation:
         PreregistrationUser.objects.filter(
-            email__iexact=user_profile.delivery_email).exclude(id=prereg_user.id)\
+            email__iexact=user_profile.delivery_email, realm=user_profile.realm).exclude(id=prereg_user.id)\
             .update(status=confirmation_settings.STATUS_REVOKED)
 
         if prereg_user.referred_by is not None:
             notify_invites_changed(user_profile)
-    else:
-        PreregistrationUser.objects.filter(email__iexact=user_profile.delivery_email)\
+    elif prereg_user is None:
+        assert not realm_creation
+        PreregistrationUser.objects.filter(email__iexact=user_profile.delivery_email, realm=user_profile.realm)\
```

Comment:
I don't understand -- with `realm_creation`, don't we want `realm=None`?

Reply:
realm_creation should never happen without a prereg_user being provided - after all we're coming from a confirmation link for realm creation. And in that case we want to activate this specific prereguser and leave all other ones alone - the email address might have requested two realm creation links because they want to create two organization. I added a commit with a test for such a scenario.

Also the `realm_creation should never happen without a prereg_user being provided ` assumption should be more clear in the extracted revoke_... function.

Reply:
I still don't understand -- what I'd expect is that `used_pregistration_user` would have `realm=None` attached to it, and thus I expect this query to do nothing:
```
def create_preregistration_user(email: str, request: HttpRequest, realm_creation: bool=False,                                        
                                password_required: bool=True, full_name: Optional[str]=None,                                         
                                full_name_validated: bool=False) -> HttpResponse:                                                    
    realm = None                                                                                                                     
    if not realm_creation:                                                                                                           
        try:                                                                                                                         
            realm = get_realm(get_subdomain(request))                                                                                
        except Realm.DoesNotExist:                                                                                                   
            pass                                                                                                                     
    return PreregistrationUser.objects.create(                                                                                       
        email=email,                                                                                                                 
        realm_creation=realm_creation,                                                                                               
        password_required=password_required,                                                                                         
        realm=realm,                                                                                                                 
        full_name=full_name,                                                                                                         
        full_name_validated=full_name_validated,                                                                                     
    )                                                                                                                                
```

Isn't the right logic here just to mutate `used_pregistration_user` in the `realm_creation` code path?

Reply:
If you print them out, `prereg_user.realm` is `None` while `created_user_profile.realm` is the new realm in the `realm_creation` code path, and I'm pretty sure we want the `None`.

Reply:
> Isn't the right logic here just to mutate `used_pregistration_user` in the `realm_creation` code path?

That's right. This prereguser was already being mutated at the beginnng of process_new_human_user. I moved that logic into the function now because that seems less confusing. With the two new refactoring commits I think things should be more readable and easier to identify as correct behavior.